File: api/path_read.go
package vaku

import (
	"errors"
	"strconv"
)

var (
	// ErrPathRead when PathRead fails.
	ErrPathRead = errors.New("path read")
	// ErrPathReadVersion when PathReadVersion fails.
	ErrPathReadVersion = errors.New("path read version")
	// ErrVaultRead when the underlying Vault API read fails.
	ErrVaultRead = errors.New("vault read")
)

// PathRead reads data at a path.
func (c *Client) PathRead(p string) (map[string]any, error) {
	vaultPath, mv, err := c.rewritePath(p, vaultRead)
	if err != nil {
		return nil, newWrapErr(p, ErrPathRead, err)
	}

	secret, err := c.vl.Read(vaultPath)
	if err != nil {
		if c.ignoreAccessErrors {
			return nil, nil
		}
		return nil, newWrapErr(p, ErrPathRead, newWrapErr(err.Error(), ErrVaultRead, nil))
	}

	if secret == nil || secret.Data == nil {
		return nil, nil
	}

	data := secret.Data
	if mv == mv2 {
		data = extractV2Read(data)
	}

	return data, nil
}

// extractV2Read returns data["data"] if the secret is not deleted or destroyed.
func extractV2Read(data map[string]any) map[string]any {
	if data == nil {
		return nil
	}

	if isDeleted(data) {
		return nil
	}

	dd := data["data"]
	if dd == nil {
		return nil
	}

	dm, ok := dd.(map[string]any)
	if !ok {
		return nil
	}

	return dm
}

// isDeleted checks if the secret has been deleted or destroyed.
func isDeleted(data map[string]any) bool {
	metadata, ok := data["metadata"].(map[string]any)
	if !ok {
		return true
	}
	deletionTime, ok := metadata["deletion_time"].(string)
	if !ok || deletionTime != "" {
		return true
	}
	destroyed, ok := metadata["destroyed"].(bool)
	if !ok || destroyed {
		return true
	}

	return false
}

// PathReadVersion reads data at a path for a specific version. Only works on v2 kv engines.
func (c *Client) PathReadVersion(p string, version int) (map[string]any, error) {
	vaultPath, mv, err := c.rewritePath(p, vaultRead)
	if err != nil {
		return nil, newWrapErr(p, ErrPathReadVersion, err)
	}

	// Only v2 mounts support versioning
	if mv != mv2 {
		return nil, newWrapErr(p, ErrPathReadVersion, ErrMountVersion)
	}

	// Pass version as query parameter using ReadWithData
	queryParams := map[string][]string{
		"version": {strconv.Itoa(version)},
	}

	secret, err := c.vl.ReadWithData(vaultPath, queryParams)
	if err != nil {
		return nil, newWrapErr(p, ErrPathReadVersion, newWrapErr(err.Error(), ErrVaultRead, nil))
	}

	if secret == nil || secret.Data == nil {
		return nil, nil
	}

	// Always extract v2 data since this function only works on v2
	return extractV2Read(secret.Data), nil
}

File: src/api/path_read.rs
//! `PathRead` and `PathReadVersion`.

use crate::api::client::Client;
use crate::api::error::{Error, ErrorKind};
use crate::api::mounts::{rewrite_path, MountVersion, VaultOp};
use crate::api::secret::{extract_v2_read, KV2_VERSION};
use serde_json::{Map, Value};

impl Client {
    /// Read the current data at a path. Returns `None` when the path doesn't
    /// exist or has been deleted/destroyed.
    pub async fn path_read(&self, p: &str) -> Result<Option<Map<String, Value>>, Error> {
        let (vault_path, mv) = rewrite_path(self.src().mount_provider.as_ref(), p, VaultOp::Read)
            .await
            .map_err(|e| Error::wrap(p, ErrorKind::PathRead, Some(Box::new(e))))?;

        let res = self.src().logical.read(&vault_path).await;
        let secret = match res {
            Ok(s) => s,
            Err(e) if self.ignore_access_errors() => {
                let _ = e;
                return Ok(None);
            }
            Err(e) => {
                return Err(Error::wrap(
                    p,
                    ErrorKind::PathRead,
                    Some(Box::new(Error::wrap(&e.to_string(), ErrorKind::VaultRead, None))),
                ))
            }
        };

        let Some(secret) = secret else { return Ok(None) };
        let Some(data) = secret.data else { return Ok(None) };

        if matches!(mv, MountVersion::Mv2) {
            Ok(extract_v2_read(Some(&data)))
        } else {
            Ok(Some(data))
        }
    }

    /// Read a specific version of a secret. KV v2 only.
    pub async fn path_read_version(
        &self,
        p: &str,
        version: i64,
    ) -> Result<Option<Map<String, Value>>, Error> {
        let (vault_path, mv) = rewrite_path(self.src().mount_provider.as_ref(), p, VaultOp::Read)
            .await
            .map_err(|e| Error::wrap(p, ErrorKind::PathReadVersion, Some(Box::new(e))))?;

        if !matches!(mv, MountVersion::Mv2) {
            return Err(Error::wrap(
                p,
                ErrorKind::PathReadVersion,
                Some(Box::new(Error::wrap("", ErrorKind::MountVersion, None))),
            ));
        }

        let version_s = version.to_string();
        let secret = self
            .src()
            .logical
            .read_with_data(&vault_path, &[(KV2_VERSION, &version_s)])
            .await
            .map_err(|e| {
                Error::wrap(
                    p,
                    ErrorKind::PathReadVersion,
                    Some(Box::new(Error::wrap(&e.to_string(), ErrorKind::VaultRead, None))),
                )
            })?;

        let Some(secret) = secret else { return Ok(None) };
        let Some(data) = secret.data else { return Ok(None) };
        Ok(extract_v2_read(Some(&data)))
    }
}
